src/Tools/Microsoft.dotnet-openapi/src/Application.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.Build.Locator;
+using Microsoft.DotNet.OpenApi.Commands;
+using Microsoft.Extensions.CommandLineUtils;
+
+namespace Microsoft.DotNet.OpenApi
+{
+    internal class Application : CommandLineApplication
+    {
+        static Application()
+        {
+            MSBuildLocator.RegisterDefaults();
+        }
+
+        public Application(
+            string workingDirectory,
+            Func<string, Task<Stream>> downloadProvider,
+            TextWriter output = null,
+            TextWriter error = null)
+        {
+            DownloadProvider = downloadProvider;
+            Out = output ?? Out;
+            Error = error ?? Error;
+
+            WorkingDirectory = workingDirectory;
+
+            Name = "microsoft.openapi";
+            FullName = "OpenApi reference management tool";
+            Description = "OpenApi reference management operations.";
+            ShortVersionGetter = GetInformationalVersion;
+
+            HelpOption("-?|-h|--help");
+
+            Invoke = () =>
+            {
+                ShowHelp();
+                return 0;
+            };
+
+            Commands.Add(new AddCommand(this));
+            Commands.Add(new RemoveCommand(this));
+            Commands.Add(new RefreshCommand(this));
+        }
+
+        public Func<string, Task<Stream>> DownloadProvider { get; }
+
+        public string WorkingDirectory { get; }
+
+        public new int Execute(params string[] args)
+        {
+            try
+            {
+                return base.Execute(args);
+            }
+            catch (AggregateException ex) when (ex.InnerException != null)
+            {
+                foreach (var innerException in ex.Flatten().InnerExceptions)
+                {
+                    Error.WriteLine(innerException.Message);
+                    if (!(innerException is ArgumentException))
+                    {
+                        Error.WriteLine(innerException.StackTrace);
+                    }
+                }
```

Comment:
I initially went along with this, but when usability testing it locally I remembered why I had done it this way in the first place.

Basically I'm using exceptions to show the users their error messages, so it really ought to be more user friendly then your actual problem being buried in amongst an Aggregate exception format and a bunch of stack-trace stuff. This all stems from me using exceptions to bail out of functions early when we find argument problems (url is null, doesn't exist) so I don't have to make every method return `Result<T>` where Result could have an error that should cause us to bail early or the value we actually asked for. If you can think of a better paradigm than using exceptions of that gross Result<T> stuff I'm on board, because I agree that using exceptions this way isn't great, it's just the best I could come up with.